http/peer-id-auth.md

```diff
@@ -0,0 +1,131 @@
+
+## Mutual Client and Server Peer ID Authentication
+
+1. The server initiates the authentication by responding to a request that must
+   be authenticated with the response header `WWW-Authenticate: Libp2p-PeerID
+   challenge-client="<base64-encoded-challenge>`. The challenge MUST be
+   indistinguishable from random data. The Server MAY randomly generate this
+   data, or MAY use an server-encrypted value. If using random data the
+   server SHOULD store the challenge temporarily until the authentication is
+   done. The challenge SHOULD be at least 32 bytes.
+
+1. The client sends a request and sets the `Authorization`
+   [header](https://www.rfc-editor.org/rfc/rfc9110.html#section-11.6.2) header
+   to the following:
+   ```
+   Libp2p-PeerID peer-id="<encoded-peer-id-bytes>",[challenge-server="<base64-encoded-challenge-server>",]sig="<base64-signature-bytes>"]
+   ```
+
+   * The `challenge-server` parameter is optional. The client should set it if
+     the client wants to authenticate the server.
+   * The peer-id is encoded per the [peer-ids spec](../peer-ids/peer-ids.md).
+   * The signature is over the concatenated result of:
+   ```
+     <varint-length> + "origin=" + server-name + 
+     [<varint-length> + "challenge-server=" + base64-encoded-client-chosen-challenge-server + ]
+     <varint-length> + "challenge-client=" + base64-encoded-challenge
+   ```
+   * Strings are UTF-8 encoded.
+   * If the challenge server was omitted in the `Authorization` header it MUST
+     be omitted in the signature.
+   * If provided, the client-chosen `challenge-server`` MUST be randomly generated.
+   * The client-chosen `challenge-server` SHOULD be at least 32 bytes.
+   * The client MUST use the same server-name as what is used for the TLS
+     session.
+   * If the client _only_ wants to authenticate the server and the server does
+     not need to authenticate the client, the client can omit the
+     `challenge-client` from the parameters and signature on its initial request
+     (since it did not receive a `challenge-client`). If a resource requires
+     client authentication, the server MUST return `401 Unauthorized` if a
+     client attempts to authenticate without a `challenge-client`.
+   * Example on building the message to sign:
+    ```
+    origin=example.com
+    client-challenge=qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=
+    challenge=AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
+    ```
+
+    Message To Sign in hex with comments
+    ```
+    12 // (18 bytes)
+    6f726967696e3d6578616d706c652e636f6d // (origin=example.com)
+
+    3d // (61 bytes)
+    636c69656e742d6368616c6c656e67653d7171717171717171717171717171717171717171717171717171717171717171717171717171717171716f3d // (client-challenge=qq...o=)
+
+    36 // (54 bytes)
+    6368616c6c656e67653d414141414141414141414141414141414141414141414141414141414141414141414141414141414141413d // (challenge=AA...A=)
+    ```
```

Comment:
This is great! I'd appreciate if we can have a more elaborate set of test vectors at the end of the document.

Reply:
I'll make a note to include this as well when I have an implementation :)